catalog/dags/data_augmentation/rekognition/add_rekognition_labels.py

```diff
@@ -0,0 +1,151 @@
+    presence of the CURRENT_POS_VAR_NAME Variable when starting the DAG. If the
+    Variable is present, assume that the DAG is resuming from an existing run,
+    otherwise assume a fresh run.
+    """
+    if Variable.get(constants.CURRENT_POS_VAR_NAME, default_var=None):
+        # Skip table creation and indexing
+        return constants.NOTIFY_RESUME_TASK_ID
+    return constants.NOTIFY_START_TASK_ID
+
+
+def _process_labels(labels: list[types.Label]) -> list[types.MachineGeneratedTag]:
+    tags = []
+    for label in labels:
+        name = label["Name"]
+        # Map name if a correction exists for it
+        name = LABEL_MAPPING.get(name, name)
+        tags.append(
+            {
+                "name": name,
+                # Confidence values need to be between 0 and 1
+                "accuracy": label["Confidence"] / 100,
+                "provider": constants.PROVIDER,
+            }
+        )
+    return tags
+
+
+def _insert_tags(tags_buffer: types.TagsBuffer, postgres_conn_id: str):
+    logger.info(f"Inserting {len(tags_buffer)} records into the temporary table")
+    postgres = PostgresHook(
+        postgres_conn_id=postgres_conn_id,
+        default_statement_timeout=constants.INSERT_TIMEOUT,
+    )
+    postgres.insert_rows(constants.TEMP_TABLE_NAME, tags_buffer, executemany=True)
+
+
+@task(trigger_rule=TriggerRule.NONE_FAILED_MIN_ONE_SUCCESS)
+def parse_and_insert_labels(
+    s3_bucket: str,
+    s3_prefix: str,
+    in_memory_buffer_size: int,
+    file_buffer_size: int,
+    postgres_conn_id: str,
+) -> types.ParseResults:
+    tags_buffer: types.TagsBuffer = []
+    failed_records = []
+    total_processed = 0
+    total_skipped = 0
+    known_offset = Variable.get(
+        constants.CURRENT_POS_VAR_NAME,
+        default_var=None,
+        deserialize_json=True,
+    )
+
+    s3_client = S3Hook(aws_conn_id=AWS_CONN_ID).get_client_type("s3")
+    with smart_open.open(
+        f"{s3_bucket}/{s3_prefix}",
+        transport_params={"buffer_size": file_buffer_size, "client": s3_client},
+    ) as file:
+        # Navigate to known offset if available
+        if known_offset:
+            logger.info(f"Previous offset found, seeking to: {known_offset}")
+            file.seek(known_offset)
+
+        # Begin parsing the file
+        # Cannot use "for blob in file" because we cannot iterate over the file
+        # and also use file.tell() to get the current position
+        while blob := file.readline():
+            total_processed += 1
+            try:
+                labeled_image: types.LabeledImage = json.loads(blob)
+            except json.JSONDecodeError:
+                logger.error(f"Failed to parse JSON: {blob}")
+                # Only track the first `n` failed records, if something
+                # is systematically wrong we don't want this getting massive
+                if len(failed_records) <= constants.MAX_FAILED_RECORDS:
+                    failed_records.append(blob)
```

Comment:
Can we continue to keep track of the _count_ of failed records? I agree we don't want to pass a ton of failed records through XCOMs, but it feels extremely important to accurately present the number that actually failed (for example, I tested locally by raising a JSONDecodeError on *every* line and the Slack message indicated to me that only 101 had failed, which is really misleading).

Only the first 5 of these even get logged to Slack (understandably!), so passing up to 100 through doesn't seem necessary either. Could we have an accurate failed_record_count, and then separately a list of up to 5 failures?

Reply:
That's a really good point! I was trying to think of _some_ way to capture things without overloading XComs. I think paired with raising an exception on systematic errors (anything over 100), capturing the failure count in the same way will be great.